File: 2015/Day9/main.py
import itertools

def parse_input(filename):
    distances = {}
    cities = set()
    with open(filename, 'r') as file:
        for line in file:
            parts = line.strip().split(' = ')
            if len(parts) == 2:
                route, distance = parts
                city1, city2 = route.split(' to ')
                distances[(city1, city2)] = int(distance)
                distances[(city2, city1)] = int(distance)
                cities.add(city1)
                cities.add(city2)
    return distances, cities

def calculate_route_distance(route, distance):
    return sum(distance[(route[i], route[i+1])] for i in range(len(route) - 1))

def find_shortest_route(distances, cities):
    shortest_distance = float('inf')
    for route in itertools.permutations(cities):
        distance = calculate_route_distance(route, distances)
        shortest_distance = min(shortest_distance, distance)
    return shortest_distance

def find_longest_route(distances, cities):
    longest_distance = float('-inf')
    for route in itertools.permutations(cities):
        distance = calculate_route_distance(route, distances)
        longest_distance = max(longest_distance, distance)
    return longest_distance

distances, cities = parse_input('input.txt')
print(f'Part 1: {find_shortest_route(distances, cities)}')
print(f'Part 2: {find_longest_route(distances, cities)}')

File: 2015/Day9/main.dart
import 'dart:io';
import 'dart:math';

typedef Distances = Map<String, int>;
typedef Cities = Set<String>;

(Distances, Cities) parseInput(String filename) {
  Distances distances = {};
  Cities cities = {};
  var lines = File(filename).readAsLinesSync();

  for (var line in lines) {
    var parts = line.split(' = ');
    if (parts.length == 2) {
      var route = parts[0];
      var distance = int.parse(parts[1]);
      var citiesPair = route.split(' to ');
      var city1 = citiesPair[0];
      var city2 = citiesPair[1];
      distances['$city1-$city2'] = distance;
      distances['$city2-$city1'] = distance;
      cities.add(city1);
      cities.add(city2);
    }
  }

  return (distances, cities);
}

int calculateRouteDistance(List<String> route, Distances distances) {
  var totalDistance = 0;
  for (var i = 0; i < route.length - 1; i++) {
    totalDistance += distances['${route[i]}-${route[i + 1]}']!;
  }
  return totalDistance;
}

int findShortestRoute(
    Distances distances, Cities cities, List<List<String>> permutations) {
  var shortestDistance = double.infinity;

  for (var route in permutations) {
    var distance = calculateRouteDistance(route, distances);
    shortestDistance = min(shortestDistance, distance.toDouble());
  }

  return shortestDistance.toInt();
}

int findLongestRoute(
    Distances distances, Cities cities, List<List<String>> permutations) {
  var longestDistance = double.negativeInfinity;

  for (var route in permutations) {
    var distance = calculateRouteDistance(route, distances);
    longestDistance = max(longestDistance, distance.toDouble());
  }

  return longestDistance.toInt();
}

List<List<T>> permute<T>(List<T> items) {
  if (items.length == 1) {
    return [items];
  }

  var perms = <List<T>>[];
  for (var i = 0; i < items.length; i++) {
    var current = items[i];
    var remaining = List<T>.from(items)..removeAt(i);
    for (var perm in permute(remaining)) {
      perms.add([current, ...perm]);
    }
  }

  return perms;
}

void main() {
  var (distances, cities) = parseInput('input.txt');
  var permutations = permute(cities.toList());

  print('Part 1: ${findShortestRoute(distances, cities, permutations)}');
  print('Part 2: ${findLongestRoute(distances, cities, permutations)}');
}
